src/components/input/StateEditor.ts

```diff
@@ -0,0 +1,72 @@
+import { State } from "./State"
+
+export class StateEditor {
+	get value(): string { return this.state.value }
+	private get selectionStart(): number { return this.state.selectionStart }
+	private set selectionStart(selectionStart: number) { this.state.selectionStart = selectionStart }
+	private get selectionEnd(): number { return this.state.selectionEnd }
+	private set selectionEnd(selectionEnd: number) { this.state.selectionEnd = selectionEnd }
+	get stateCopy(): State { return { value: this.state.value, selectionStart: this.selectionStart, selectionEnd: this.selectionEnd}}
+
+	private constructor(private state: State) { }
+
+	get(index: number): string {
+		return this.value[index]
+	}
+	isDigit(index: number): boolean {
+		const character = this.get(index)
+		return character >= "0" && character <= "9"
+	}
+
+	match(matcher: { [Symbol.match](string: string): RegExpMatchArray | null }): RegExpMatchArray | null {
+		return this.value.match(matcher)
+	}
+	replace(value: string, start: number, end: number): StateEditor {
+		this.state.value = this.value.slice(0, start) + value + this.value.slice(end, this.value.length)
+		if (this.selectionStart >= end)
+			this.selectionStart = this.selectionStart + value.length - end + start
+		else if (this.selectionStart > start && this.selectionStart < end)
+			this.selectionStart = Math.min(this.selectionStart, start + value.length)
+		if (this.selectionEnd >= end)
+			this.selectionEnd = this.selectionEnd + value.length - end + start
+		else if (this.selectionEnd > start && this.selectionEnd < end)
+			this.selectionEnd = Math.min(this.selectionEnd, start + value.length)
+		return this
+	}
+	insert(value: string, index: number): StateEditor {
+		return this.replace(value, index, index)
+	}
+	delete(start: number, end?: number): StateEditor {
+		return this.replace("", start, end || start + 1)
+	}
+	truncate(end: number): StateEditor {
+		if (this.value.length >= end) {
+			this.delete(end, this.value.length)
+		}
+		return this
+	}
+	private pad(length: number, padding: string, padStart: boolean): StateEditor {
```

Comment:
Replace `padStart: boolean` with `index: number` instead. If we figure out a suitable name we can even expose it publicly.

Reply:
padAtIndex?

Reply:
`padAt` is enough, but just `pad` will also work.